### PR TITLE
Matcha 18 add tracing time limitations

### PR DESCRIPTION
**Pull request by OnorioCatenacci**
I'm not sure if this is what you were looking for on this issue Chris but I figured I'd take a whack at it anyway.  I didn't want to invest too much effort (I mean in updating docs and such) since I wasn't sure if it was what you wanted or not.  My developer testing worked and none of the unit tests etc. failed but I realize that doesn't prove it's not a problem. 

I did get a message from Credo regarding cyclomatic complexity in do_recon_trace_calls and I suspect that's because I added a new if statement into the function.  I'm not sure how else I could address that one though; I mean I think I need to see if the user specified a duration (rather than it coming through as the default) and if they did not then I shouldn't start a process timer going. 

At any rate, I hope this helps Chris!

### CODE REVIEW

**Comment by OnorioCatenacci** [comment]
I think I've dealt with the failing test @christhekeele if you want to give this a look when you've got time.

**Comment by OnorioCatenacci** [comment]
Further testing has shown me that I don't have this fixed correctly yet.  I am closing this PR until I can get it into better shape.

File: lib/matcha/trace.ex [hunk 1]
```diff
@@ -17,6 +17,7 @@ defmodule Matcha.Trace do
 
   @default_trace_limit 1
   @default_trace_pids :all
+  @default_duration :indefinite
```
Comment (christhekeele) [inline lib/matcha/trace.ex:20]: There's a venerable tradition of using the atom `:infinity` in Erlang and Elixir for infinite time durations, especially timeouts. I think that idiom suits this situation well.

**Review by christhekeele — COMMENTED** [review]
This does help, jumpstarting the conversation if nothing else!

### Quick Notes

- Re: credo, I don't care about the cyclomatic complexity stuff too much, I change the threshold as I find complex functions that I think are actually fine. We can do that for this or any PR.

- Re: tests, it looks like you fixed most of the typecheck failures, feel free to ask me about any other failures in particular if it's not clear what's going on!

### Implementation

I like the approach you're taking here, working outside of recon (which does not support this in its [trace calls options](https://ferd.github.io/recon/recon_trace.html#t:options/0)). Set a timer to send a message, and handle the message asynchronously to stop the trace.

I think I see why it's not working as writ, though. We're [using `Process.send_after(self())`](https://github.com/christhekeele/matcha/pull/62/files#diff-d0864196372507018bc07ea43e58cb6b5f0b76e0a050a6a7d102e3229827b0d6R112) to send a message to the calling process. However, we'd need the caller process to be checking its own mailbox with a `receive` loop to act on it. `GenServer`s wrap this idiom with a `handle_info` callback. However, we can't assume the caller is a `GenServer`, and even if it was, that `GenServer` would have to manually call back into the [`Matcha` function you have written](https://github.com/christhekeele/matcha/pull/62/files#diff-d0864196372507018bc07ea43e58cb6b5f0b76e0a050a6a7d102e3229827b0d6R460-R462) for it to actually fire.

The real challenge here is that traces are bound to the calling process: the erlang tracing engine will send trace messages to the caller that initiated the trace, and only that process can tell the engine that it wants to stop receiving trace messages (without hacks). But, we do not control the calling process, so cannot issue this instruction to the tracing engine ourselves from within the caller ourselves.

`recon` works around this by running [its own `GenServer` architecture](https://ferd.github.io/recon/recon_trace.html#module-structure), starting the trace there, and proxying results back to the caller process. It implements its own `:limit` (and not-yet-documented-in-`Matcha` `:rate_limit`) throttling mechanisms within that `GenServer`.

So, if we want to extend this with new limitation mechanisms, we'd need a similar setup. That setup could then proxy to `recon`, which runs a similar setup... At which point, we might as well use our own setup to do what `recon` does ourselves, which would be #3. So perhaps we should do that first?

**Comment by OnorioCatenacci** [comment]
> This does help, jumpstarting the conversation if nothing else!
> 
> ### Quick Notes
> * Re: credo, I don't care about the cyclomatic complexity stuff too much, I change the threshold as I find complex functions that I think are actually fine. We can do that for this or any PR.
> * Re: tests, it looks like you fixed most of the typecheck failures, feel free to ask me about any other failures in particular if it's not clear what's going on!
> 
> ### Implementation
> I like the approach you're taking here, working outside of recon (which does not support this in its [trace calls options](https://ferd.github.io/recon/recon_trace.html#t:options/0)). Set a timer to send a message, and handle the message asynchronously to stop the trace.
> 
> I think I see why it's not working as writ, though. We're [using `Process.send_after(self())`](https://github.com/christhekeele/matcha/pull/62/files#diff-d0864196372507018bc07ea43e58cb6b5f0b76e0a050a6a7d102e3229827b0d6R112) to send a message to the calling process. However, we'd need the caller process to be checking its own mailbox with a `receive` loop to act on it. `GenServer`s wrap this idiom with a `handle_info` callback. However, we can't assume the caller is a `GenServer`, and even if it was, that `GenServer` would have to manually call back into the [`Matcha` function you have written](https://github.com/christhekeele/matcha/pull/62/files#diff-d0864196372507018bc07ea43e58cb6b5f0b76e0a050a6a7d102e3229827b0d6R460-R462) for it to actually fire.
> 
> The real challenge here is that traces are bound to the calling process: the erlang tracing engine will send trace messages to the caller that initiated the trace, and only that process can tell the engine that it wants to stop receiving trace messages (without hacks). But, we do not control the calling process, so cannot issue this instruction to the tracing engine ourselves from within the caller ourselves.
> 
> `recon` works around this by running [its own `GenServer` architecture](https://ferd.github.io/recon/recon_trace.html#module-structure), starting the trace there, and proxying results back to the caller process. It implements its own `:limit` (and not-yet-documented-in-`Matcha` `:rate_limit`) throttling mechanisms within that `GenServer`.
> 
> So, if we want to extend this with new limitation mechanisms, we'd need a similar setup. That setup could then proxy to `recon`, which runs a similar setup... At which point, we might as well use our own setup to do what `recon` does ourselves, which would be #3. So perhaps we should do that first?

Agreed about tackling #3 first.  In fact I've got a Matcha_3 branch here on my local machine but I have to confess my Erlang reading skills aren't quite up to snuff because I can't quite follow what he's done in the recon code.  I mean I got it working (sorta) with a direct call to :erlang.trace/3 but I'm not confident that it's the best approach to making it work right.  

I didn't know about the :infinity atom; thanks for pointing that one out to me.  

Tell you what I think I'll do; I'll share the work I've done on #3 (incomplete though it is) and maybe you can help me to iron out the things that aren't there yet.

**Comment by christhekeele** [comment]
I've been procrastinating reading recon's erlang too, as I'm not particularly fluent 🤣. I'll sit down and eat my vegetables when looking over your exploration!

**Comment by christhekeele** [comment]
> I didn't know about the `:infinity` atom; thanks for pointing that one out to me.

Yep! The most place people first encounter this is [in the `GenServer` timeout docs](https://hexdocs.pm/elixir/GenServer.html#module-timeouts), but it's pretty common in other places in the erlang stdlib too!

**Pull request closed without merging**